Document the syntax changes introduced by CVL 2 (#57)
Certora/Documentation: docs/cvl/cvl2/changes.md

Syntax changes introduced in CVL 2
==================================

> CVL 2.0 is a major overhaul to the type system of CVL.  Many

Reviewer: I'd argue it's gone beyond the type system. I would say it's an overhaul to the "internals" and leave it at that.

Reviewer: oh you literally say that on the next line. well, never mind then, leave as is

introduce a few improvements to the syntax.  The general goal of these changes
is to make the behavior of CVL more explicit and predictable, and to bring the
syntax more in line with solidity's syntax.

This document summarizes the changes to CVL syntax introduced by CVL 2.0.

```{contents}
```

Superficial syntax changes
--------------------------

There are several simple changes to the syntax to make specs more uniform and
consistent, and to reduce the superficial differences with Solidity.

### `function` and `;` required for methods block entries
Methods block entries must now start with `function` and end with `;`.  For
example:

```cvl
balanceOf(address) returns(uint) envfree
```
will become
```cvl
function balanceOf(address) external returns(uint) envfree;
```
(note also the addition of `external`, see {ref}`described below <cvl2-visibility>`).

This is also true for entries with summaries:
```cvl
_setManagedBalance(address,uint256) => NONDET
```
will become
```cvl
function _setManagedBalance(address,uint256) internal => NONDET;
```

```{todo}
If you do not change this, you will see the following error:
```

### Required `;` in more places

> `using`, `pragma`, `import`, and `use` statements all require a `;` at the end.  For

Reviewer 2: `;` if there's no block, no `;` if there is a block

> `using`, `pragma`, `import`, and `use` statements all require a `;` at the end.  For

Reviewer 3: Right now you can't put a semicolon also when there is no block.

would become:

```cvl
require f.selector == sig:balanceOf(address).selector;
```

```{todo}
If you do not change this, you will see the following error:
```

### Stricter ordering on method annotations

In CVL 2, the order of the annotations must be visibility modifiers (`internal` or `external`),
followed by `returns` clause (if any), followed by `optional`, `library`, or `envfree` in any order (if any),
followed by a summary (if any).

CVL 1 was less strict about the order.

```{todo}
If you do not change this, you will see the following error:
```

### Use of contract name instead of `using` variable

In CVL 1, the only way to refer to a contract on the scene was to first
introduce a variable with a `using` statement, and then use that variable.  For
example, to access a struct type `S` defined in `Example.sol`, you would need
to write

```cvl
using Example as c;

rule example {
    c.S x = getAnS();
}
```

In CVL 2, you must now use the name of the contract, rather than the variable,
when referring to user-defined types.  The above example would now be written

```cvl
rule example {
    Example.S x = getAnS();
}
```

There is no need for a `using` statement in this example.

> `using` statements are still required to call methods on secondary contracts.

Author: Do you mean pre- or post- CVL 2?  This document is intended to be the CVL 2 documentation.

```{todo}
Error message
```

Changes to methods block entries

> --------------------------------

Reviewer 3: We need to add to this section that user-defined types (enum, user-defined-value-type, and struct) now _must_ be written with the user-defined names. i.e. uint8 can't replace enums and tuples can't replace structs (I think in CVL1 you _had_ to use tuples?)

In addition to the superficial changes listed above, there are some changes that
change the way that methods block entries can be written.  In CVL 1, `methods`
block entries often had several different functions and meanings:

 - They are used to indicate targets for summarization
 - They are used to write generic specs that could apply to contracts with
   missing methods
 - They are used to declare targets `envfree`

With these changes, these different uses are more explicit.

(cvl2-visibility)=
### Required `internal` or `external` annotation

Every methods block entry must be marked either `internal` or `external`.  The
annotation must come after the argument list and before the `returns` clause.

If a function is declared `public` in Solidity, then the Solidity compiler
creates an internal implementation method, and an external wrapper method that
calls the internal implementation.  Therefore, you can summarize a `public`
method by marking the summarization `internal`.

> If you want to summarize both the internal implementation and the external

Reviewer 2: This may or may not be the place to clarify that if the summaries are different, the `external` summary declaration will (I _believe_) take precedence, though if the `external` declaration has no summary, the internal summary will still be applied.

There is probably a better way to describe this but hopefully you get the drift

Author: I added a more extensive description

wrapper, you need to add two separate entries to the `methods` block.

```{todo}
If you do not change this, you will see the following error:
```

### `optional` methods entries

In CVL 1, you could write an entry in the methods block for a method that does

> not exist in the contract; rules that would call the non-existent method are

Reviewer 2: "were skipped"? "would be skipped"?

skipped during verification.

This behavior can lead to confusion, because typos or name changes could silently
cause a rule to be skipped.

In CVL 2, this behavior is still available, but the methods entry must contain
the keyword `optional` somewhere after the `returns` clause and before the
summarization (if any).

```{todo}
If a methods block contains a non-optional entry for a method that doesn't exist
in the contract, you will receive the following error message:
```

### `library` annotations

> In CVL 2, contract functions declared as library functions must be annotated

Reviewer: no, I think we pretty rarely summarized external library functions

> In CVL 2, contract functions declared as library functions must be annotated

Author: I created https://certora.atlassian.net/browse/CERT-1747

with `library` in the `methods` block.

```{todo}
If you forget to declare a method as a `library` method, you will receive the
following error message:
```

### Required `calldata`, `memory`, or `storage` annotations for reference types

In CVL 2, methods entries for internal functions must contain either `calldata`,
`memory`, or `storage` annotations for all arguments with reference types (such
as arrays).

```{todo}

> is `calldata` actually one of the options?

Reviewer 2: It is definitely possible to write an internal function with a parameter in a `calldata` location

```

```{todo}
If you do not change this, you will see the following error:
```

### Summaries only apply to one contract by default

In CVL 1, a summary in the `methods` block applied to all methods with the
given signature.  Entries that had both an explicit receiver and a summary,

> such as the following, were disallowed:

Reviewer 2: I don't think this was true? We definitely had multi-contract summaries

the default is `currentContract`.

Consider the following example:
```cvl
using C as c;

methods {
    function f(uint)   => NONDET;
    function c.g(uint) => ALWAYS(4);
    function h(uint)   => ALWAYS(1);
    function _.h(uint) => NONDET;
}
```

In this example, `currentContract.f` has a `NONDET` summary, `c.g` has an `ALWAYS`
summary, a call to `currentContact.h` has an `ALWAYS` summary and a call to
`h(uint)` on any other contract will use a `NONDET` summary.

Summaries for specific contract methods (including the default
`currentContract`) always override wildcard summaries.

Wildcard entries cannot be declared `optional` or `envfree`, since these
annotations only make sense for specific contract methods.

```{warning}
The meaning of your summarizations will change from CVL 1 to CVL 2.  In CVL 2,
any entry without a `_` will only apply to a single contract!
```

### Requirements on `returns`

In CVL 2, methods block entries require a `returns` clause in the following
situations:

 - The method is not a wildcard entry and the contract function returns a value.
   In this case, the methods block entry must match the contract function's

>    return type.

Reviewer: wait really? you can't just do `foo(uint) envfree`?

Reviewer: Also, I'm not sure what this means. If the contract function returns something you need to have a return annotation. So, isn't that always?

Author: I don't think you need to (or can) explicitly indicate that something void?

   ```{todo}
   Error message
   ```

 - The method is a wildcard entry that is summarized with a ghost or CVL
   function summary.  In this case, the return type from the CVL function must

>    be compatible with the declared return type for the method.

Author: I added a big warning

   ```{todo}
   Error message
   ```

> In all other situations, a `returns` clause is forbidden.

Reviewer 2: This makes it sound like a lot of situations. Maybe explicitly remind that this is for wildcard, non ghost/cvl function summaries

```{todo}
If you do not change this, you will see the following error:
```

Changes to integer types
------------------------

In CVL 1, the rules for casting between integer types were complex; CVL 2
simplifies them.

> The general rule of thumb is that you should use `mathint` for all function

Reviewer 2: This is a bit confusing. It is impossible to use a `mathint` for a contract function output. Also, why would we not want `mathints` as CVL Function inputs or ghost function inputs?

It is now impossible for CVL math expressions to cause overflow - all integer
operations are exact.

### Mathematical operations return `mathint`

In CVL 2, the result of all arithmetic operators have type `mathint`,
regardless of the input types.  Arithmetic operators include `+`,
`*`, `-`, `/`, `^`, and `%`, but not bitwise operators like `<<`, `xor`, and `|`
(changes to bitwise operators are described {ref}`below <cvl2-bitwise>`).

The primary impact of this change is that you may need to declare more of your
variables as `mathint` instead of `uint`.  If you are performing arithmetic
operations and integers that you are passing to specs, you will need to be more
explicit about the overflow behavior by using the {ref}`new casting operators
<cvl2-casting>`.

```{todo}
If you do not change this, you will see the following error:
```

### Comparisons require identical types

When comparing two integers using `==`, `<=`, `<`, `>`, or `>=`, CVL 2 will
require both sides of the equation to have identical types, and {ref}`implicit
casts <cvl2-casting>` will not be used.  Comparisons with number literals (e.g.
`0` or `1`) are allowed for any integer type.

> If you do not have identical types, the best solution is to use the special

Reviewer: agreed.

> If you do not have identical types, the best solution is to use the special

Reviewer 2: idk, I think this unsoundness is really a super edge case and like, our sanity checks should really catch this

> If you do not have identical types, the best solution is to use the special

Author: Do you think the example just below this is adequate?

> If you do not have identical types, the best solution is to use the special

Reviewer 4: yes, but after fixing tons of specs I think we can use the following 2 recommendations in the guide:

1. It's fine to use `to_mathint` on a value originating from Solidity, as the original container (Solidity) is bound to 256 bits anyway. 
2. Spec values such as local variables and ghosts should not be upcasted with `to_mathint`, but rather be declared as `mathint`. An exception can be made for local variables passed as arguments into Solidity code.

`to_mathint` operator to convert both sides to `mathint`.  For example:

```cvl
assert to_mathint(balanceOf(user)) == initial + deposit;
```

Note that in this example, we do not need to cast the right hand side, since
the result of `+` is always of type `mathint`.

````{todo}

> When should you not simply cast to `mathint`?  We have one example: consider the

Reviewer: wait, is this a todo? I say, do! this is good documentation (without the editorializing, obviously)

Author: The reason it's left as a TODO is because it's not clear how to generalize from this example.  Are there other situations?  What's the general rule?

following code:

```cvl
ghost uint256 sum;

hook ... {
    havoc sum assuming sum@new == sum@old + newBalance - oldBalance;
}
```

Simply casting to `mathint` will turn overflows into vacuity.  It's not clear
how to generalize from this example.

In this particular example, the right solution is to declare `sum` to be a
`mathint` instead of a `uint`.  Note that with the more "modern" update syntax,
this isn't a problem:

```cvl
ghost uint256 sum;

hook ... {
    sum = sum + newBalance - oldBalance;
}
```
will say that the right-hand side is a `mathint` which can't be assigned to a
`uint`.

> We hope that security engineers will think carefully about these changes and let

Reviewer 4: ```suggestion
We hope that specification writers will think carefully about these changes and let
```

Author: I'm hoping to have this TODO resolved during the SE test fest, before we release CVL 2 to the general public.

`2^8-1`, so `uint8` is also a subtype of `int16`.

All integer types are subtypes of `mathint`, since any integer can be
represented by a `mathint`.

In CVL 1, the rules for converting between supertypes and subtypes were
complicated; they depended not only on the types involved, but on the context
in which the conversion happened.  CVL 2 simplifies these rules and improves the
clarity and predictability of casts.

In CVL 2, with one exception, you can always use a subtype whenever the
supertype is accepted.  For example, you can always use a `uint8` where an
`int16` is expected.  We say that the subtype can be "implicitly cast" to the
supertype.

The one exception is comparison operators; as mentioned above, you must add an
explicit conversion if you want to compare two numbers with different types.
The `to_mathint` operator exists solely for this purpose; in all other contexts
you can simply use any number when a `mathint` is expected (since all integer
types are subtypes of `mathint`).

In order to convert from a supertype to a subtype, you must use an explicit
cast.  In CVL 1, the syntax for casting to a subtype was `to_<subtype>(value)`,

> for example `to_uint256(x)`.

Reviewer 4: We should mention those are no longer available in CVL2

Reviewer: and I think **only** `to_uint256` existed, other versions didn't

In CVL 2, there are now two casting operators: `assert_<type>(value)` and
`require_<type>(value)` (for example: `assert_uint8(x)` or `require_uint8(x)`).
Each of these cases checks that the value is in range; the `assert` cast will
report a counterexample if the value is out of range, while the `require` cast
will ignore counterexamples where the cast value is out of range.

```{todo}
is it an error to use an explicit cast for an upcast?
```

```{todo}
If you do not change this, you will see the following error:
```

### Modulo operator `%` returns negative values for negative inputs

As in Solidity, if `n < 0` then `n % k == -(-n % k)`.

### Support for `bytes1`...`bytes32`

CVL 2 supports the types `bytes1`, `bytes2`, ..., `bytes32`, as in Solidity.
Number literals must be explicitly cast to these types using `to_bytesN`; for
example:

```cvl
bytes32 x = to_bytes32(0);
```

> There is no way to convert between these types and integer types (except for

Reviewer 4: the crowd will ask: does `to_mathint` not work on bytesN?

Reviewer: bytes32 are described in solidity as "Fixed sized byte arrays" the fact that they are convertible to numbers is super weird imo.

Reviewer: no need to repeat solidity's mistakes

literals as just mentioned).

```{todo}
Update this if we add casting between `uint256` and `bytes32`.
```

```{todo}
If you do not change this, you will see the following error:
```

(cvl2-bitwise)=
### Changes for bitwise operations

> The exact details for bitwise operations (such as `&`, `|`, and `<<`) were not

Reviewer 4: ```suggestion
In CVL1, the exact details for bitwise operations (such as `&`, `|`, and `<<`) were not
```

completely specified, especially for negative integers.

In CVL 2, all bitwise operations (`&`, `|`, `~`, `>>`, `>>>`, `<<`, and `xor`)

> first convert to `uint256`, then perform the operations on the full 256-bit

Reviewer 2: maybe instead of saying they convert to `uint256` just say they are converted to a 256-bit word? only because signed ints aren't really converted to uints, we just get a different representation. I suppose you could just say both types use two's complement encoding, because two's complement of a uint is just, it's current encoding?

word, then convert back to the expected type.  Signed integer types use
twos-complement encoding.

The two right-shifts differ in how they treat signed integers.  `>>` is an
arithmetic shift; it preserves the sign bit.  `>>>` is a logical shift; it pads
the shifted word with zero.

> Bitwise operations cannot be performed on `mathint` values.

Reviewer 2: maybe also note that the current semantics of the underlying bitwise operations are still gross overapproximations?

### Conversion between `bytes<k>`, `address`, and integer types

```{todo}
finish
```

Removed features
----------------

As part of the transition to CVL 2.0, we have removed several language features
that are no longer used.

We have removed these features because we think they are no longer used and no
longer useful.  If you find that you do need one of these features, contact
Certora support.

### Methods entries for sighashes

In CVL 1, you could write a sighash instead of a method identifier in the
`methods` block.  This feature is no longer supported.  You will need to have
the name and argument types of the called method in order to provide an entry.

```{todo}
If you do not change this, you will see the following error:
```

### `invoke`, `sinvoke`, and `call`

Older versions of CVL had special syntax for calling contract functions:
 - `invoke f(args);` should be replaced with `f(args);`.
 - `sinvoke f(args);` should be replaced with `f@withrevert(args);`

>  - `call f(args)` should be replaced with `f(args)`.

Reviewer 2: this was not for contract functions, it was for CVL Functions (formerly called "subroutines")

```{todo}
If you do not change this, you will see the following error:
```

### `static_assert` and `static_require`

These deprecated aliases for `assert` and `require` are being removed; replace

> them with `assert` and `require`.

Reviewer 3: ```suggestion
them with `assert` and `require`, respectively.
```

```{todo}
If you do not change this, you will see the following error:
```

### `invoke_fallback`

The `invoke_fallback` syntax is no longer supported; there is no longer a way
to directly invoke the fallback method.

```{todo}
If you do not change this, you will see the following error:
```

### Havocing `calldataarg` variables

In CVL 1, you could write the following:

```cvl
calldataarg args; env e;
f(e, args);

> havoc args;

Reviewer: I think we've decided to remove havocing for everything? am I remembering incorrectly?

Reviewer 2: Yeah, the story here is probably, "you may no longer havoc local variables"

g(e, args);
```

You can no longer write `havoc x` where `x` is any variable of type `calldataarg`.

Instead, replace the havoced variable with a new variable.

```{todo}
If you do not change this, you will see the following error:
```

### Destructuring syntax for struct returns

In CVL 1, if a contract function returned a struct, you could use a
destructuring syntax to get the return value in your spec.  For example,
consider the following contract:

```solidity
contract Example {
    struct S {
        uint firstField;
        uint secondField;
        bool thirdField;
    }

    function f() returns(S) { ... }
}
```

To access the return value of `f` in CVL 1, you could write the following:

```cvl
uint x; uint y; bool z;
x, y, z = f();
```

This syntax is no longer supported.  Instead, you should declare a variable with
the struct type:

```cvl
Example.S result = f();
uint x = result.firstField;
```

> 

Reviewer 2: maybe make a note that destructuring is still supported for tuple returns

```{todo}
If you do not change this, you will see the following error:
```

### `bytes[]` and `string[]`

In CVL 1, you could declare variables of type `string[]` and `bytes[]`.  You can
no longer use these types in CVL.

You can still declare contract methods that use these types in the `methods`
block.  However, you can only call methods that take one of these types as an
argument by passing a `calldataarg` variable, and you cannot access the return
value of a method that returns one of these types.

```{todo}

> Determine whether you can call with `_`.

Reviewer 2: Very likely not. I believe the mechanism is to infer the type and generate a variable of the type internally.

This document summarizes the changes to CVL syntax introduced by CVL 2.0.

```{contents}
```

Superficial syntax changes
--------------------------

There are several simple changes to the syntax to make specs more uniform and
consistent, and to reduce the superficial differences with Solidity.

### `function` and `;` required for methods block entries
Methods block entries must now start with `function` and end with `;`.  For
example:

```cvl
balanceOf(address) returns(uint) envfree
```
will become
```cvl
function balanceOf(address) external returns(uint) envfree;
```
(note also the addition of `external`, see {ref}`described below <cvl2-visibility>`).

This is also true for entries with summaries:
```cvl
_setManagedBalance(address,uint256) => NONDET
```
will become
```cvl
function _setManagedBalance(address,uint256) internal => NONDET;
```

```{todo}
If you do not change this, you will see the following error:
```

### Required `;` in more places

`using`, `pragma`, `import`, and `use` statements all require a `;` at the end.  For
example,

```cvl
using C as c
```

becomes
```cvl
using C as c;
```

Note: `use` statements do not require (and may not have) a semicolon if they

> are followed by a `preserved` or `filtered` block.

Reviewer 5: ```suggestion
are followed by a `preserved` or `filtered` block. For example,

```cvl
use rule tree_isImmutable filtered {
    f -> !excludeFromProver(f)
}
```text to allow backticks to show up in github suggestion
```

```{todo}
If you do not change this, you will see the following error:
```

### Method literals require `sig:`

In some places in CVL, you can refer to a contract method by its name and
argument types.  For example, you might write
```cvl
require f.selector == balanceOf(address).selector;
```

In this example, `balanceOf(address)` is a *method literal*.  In CVL 2,
these methods literals must now start with `sig:`.  For example, the above
would become:

```cvl
require f.selector == sig:balanceOf(address).selector;
```

```{todo}
If you do not change this, you will see the following error:
```

### Stricter ordering on method annotations

In CVL 2, the order of the annotations must be visibility modifiers (`internal` or `external`),
followed by `returns` clause (if any), followed by `optional`, `library`, or `envfree` in any order (if any),
followed by a summary (if any).

CVL 1 was less strict about the order.

```{todo}
If you do not change this, you will see the following error:
```

### Use of contract name instead of `using` variable

> In CVL 1, the only way to refer to a contract on the scene was to first

Reviewer 5: I think "on the scene" is a technical term here rather than just a usage of the normal English colloquialism, yes? I was initially confused, although that may be my lack of familiarity with this usage.

Author: added a link to the glossary

introduce a variable with a `using` statement, and then use that variable.  For
example, to access a struct type `S` defined in `Example.sol`, you would need
to write

```cvl
using Example as c;

rule example {
    c.S x = getAnS();
}
```

In CVL 2, you must now use the name of the contract, rather than the variable,
when referring to user-defined types.  The above example would now be written

```cvl
rule example {
    Example.S x = getAnS();
}
```

There is no need for a `using` statement in this example.

`using` statements are still required to call methods on secondary contracts.

```{todo}
Error message
```

Changes to methods block entries
--------------------------------

In addition to the superficial changes listed above, there are some changes that
change the way that methods block entries can be written.  In CVL 1, `methods`

> block entries often had several different functions and meanings:

Reviewer 5: ```suggestion
block entries often have several different functions and meanings:
```
Changed to present tense since CVL 1 still exists and the points immediately below are in present tense.

Author: I decided to change them all to past tense because I think that's what the rest of the document does

 - They are used to indicate targets for summarization
 - They are used to write generic specs that could apply to contracts with
   missing methods
 - They are used to declare targets `envfree`

> With these changes, these different uses are more explicit.

Reviewer 5: This sentence is a bit vague. Perhaps it would be clearer to say something like, "In CVL 2, these different uses are more explicit." Not sure on the best wording.

wrapper, you need to add two separate entries to the `methods` block.

```{todo}
If you do not change this, you will see the following error:
```

### `optional` methods entries

In CVL 1, you could write an entry in the methods block for a method that does
not exist in the contract; rules that would call the non-existent method are
skipped during verification.

This behavior can lead to confusion, because typos or name changes could silently
cause a rule to be skipped.

In CVL 2, this behavior is still available, but the methods entry must contain
the keyword `optional` somewhere after the `returns` clause and before the
summarization (if any).

```{todo}
If a methods block contains a non-optional entry for a method that doesn't exist
in the contract, you will receive the following error message:
```

### `library` annotations

In CVL 2, contract functions declared as library functions must be annotated
with `library` in the `methods` block.

```{todo}
If you forget to declare a method as a `library` method, you will receive the
following error message:
```

### Required `calldata`, `memory`, or `storage` annotations for reference types

> In CVL 2, methods entries for internal functions must contain either `calldata`,

Reviewer 5: ```suggestion
In CVL 2, methods block entries for internal functions must contain either `calldata`,
```

`h(uint)` on any other contract will use a `NONDET` summary.

Summaries for specific contract methods (including the default
`currentContract`) always override wildcard summaries.

Wildcard entries cannot be declared `optional` or `envfree`, since these
annotations only make sense for specific contract methods.

```{warning}
The meaning of your summarizations will change from CVL 1 to CVL 2.  In CVL 2,
any entry without a `_` will only apply to a single contract!
```

### Requirements on `returns`

In CVL 2, methods block entries require a `returns` clause in the following
situations:

 - The method is not a wildcard entry and the contract function returns a value.
   In this case, the methods block entry must match the contract function's
   return type.

   ```{todo}
   Error message
   ```

 - The method is a wildcard entry that is summarized with a ghost or CVL
   function summary.  In this case, the return type from the CVL function must
   be compatible with the declared return type for the method.

   ```{todo}
   Error message
   ```

In all other situations, a `returns` clause is forbidden.

```{todo}
If you do not change this, you will see the following error:
```

> In particular, one cannot specify return types for wildcard entries, as different 

Reviewer 5: The posted build of the docs doesn't seem to include this sentence and readthedocs is showing a build failure error https://readthedocs.com/projects/certora-certora-prover-documentation/builds/1382001/

contracts could declare the same method signature with different return types.

Changes to integer types
------------------------

In CVL 1, the rules for casting between integer types were complex; CVL 2
simplifies them.

The general rule of thumb is that you should use `mathint` for all function
outputs, and the appropriate `int` or `uint` type for all function inputs.

It is now impossible for CVL math expressions to cause overflow - all integer
operations are exact.

### Mathematical operations return `mathint`

> In CVL 2, the result of all arithmetic operators have type `mathint`,

Reviewer 5: ```suggestion
In CVL 2, the results of all arithmetic operators have type `mathint`,
```

`methods` block.  This feature is no longer supported.  You will need to have
the name and argument types of the called method in order to provide an entry.

```{todo}
If you do not change this, you will see the following error:
```

### `invoke`, `sinvoke`, and `call`

Older versions of CVL had special syntax for calling contract functions:
 - `invoke f(args);` should be replaced with `f(args);`.
 - `sinvoke f(args);` should be replaced with `f@withrevert(args);`
 - `call f(args)` should be replaced with `f(args)`.

```{todo}
If you do not change this, you will see the following error:
```

### `static_assert` and `static_require`

These deprecated aliases for `assert` and `require` are being removed; replace
them with `assert` and `require`.

```{todo}
If you do not change this, you will see the following error:
```

### `invoke_fallback`

The `invoke_fallback` syntax is no longer supported; there is no longer a way
to directly invoke the fallback method.

```{todo}
If you do not change this, you will see the following error:
```

### Havocing `calldataarg` variables

In CVL 1, you could write the following:

```cvl

> calldataarg args; env e;

Reviewer 5: Syntax highlighting does not recognize `calldataarg`

Author: This is a known problem that should be addressed elsewhere https://certora.atlassian.net/browse/DOC-214?atlOrigin=eyJpIjoiZWJiZWMyYTg5YmUyNDAzNTllMWE4NWI1ZWYxNDlkOWMiLCJwIjoiaiJ9

syntax more in line with Solidity's syntax.

This document summarizes the changes to CVL syntax introduced by CVL 2.0.

```{contents}
```

Superficial syntax changes
--------------------------

There are several simple changes to the syntax to make specs more uniform and
consistent, and to reduce the superficial differences with Solidity.

### `function` and `;` required for methods block entries
Methods block entries must now start with `function` and end with `;`.  For
example:

```cvl
balanceOf(address) returns(uint) envfree
```
will become
```cvl
function balanceOf(address) external returns(uint) envfree;
```
(note also the addition of `external`, {ref}`described below <cvl2-visibility>`).

This is also true for entries with summaries:
```cvl
_setManagedBalance(address,uint256) => NONDET
```
will become
```cvl
function _setManagedBalance(address,uint256) internal => NONDET;
```

```{todo}
If you do not change this, you will see the following error:
```

### Required `;` in more places

`using`, `pragma`, `import`, and `use` statements all require a `;` at the end.  For
example,

```cvl
using C as c
```

becomes
```cvl
using C as c;
```

> Note: `use` statements do not require (and may not have) a semicolon if they

Author: ```suggestion
`use` statements do not require (and may not have) a semicolon if they
```

followed by a summary (if any).

CVL 1 was less strict about the order.

```{todo}
If you do not change this, you will see the following error:
```

### Use of contract name instead of `using` variable

In CVL 1, the only way to refer to a contract on the {term}`scene` was to first
introduce a variable with a `using` statement, and then use that variable.  For
example, to access a struct type `S` defined in `Example.sol`, you would need
to write

```cvl
using Example as c;

rule example {
    c.S x = getAnS();
}
```

In CVL 2, you must now use the name of the contract, rather than the variable,
when referring to user-defined types.  The above example would now be written

```cvl
rule example {
    Example.S x = getAnS();
}
```

There is no need for a `using` statement in this example.

`using` statements are still required to call methods on secondary contracts:

```cvl
using Example as c;

rule example {
    ...
    c.balanceOf(a);
    ...
}
```

Entries in the `methods` block may use either the contract name or the instance
variable.

```{todo}
Error message
```

Changes to methods block entries
--------------------------------

> In addition to the superficial changes listed above, there are some changes that

Reviewer 5: ```suggestion
In addition to the superficial changes listed above, there are some changes to
```

> change the way that methods block entries can be written.  In CVL 1, `methods`

Reviewer 5: ```suggestion
the way that methods block entries can be written.  In CVL 1, `methods`
```

block entries often had several different functions and meanings:

 - They were used to indicate targets for summarization
 - They were used to write generic specs that could apply to contracts with
   missing methods
 - They were used to declare targets `envfree`

The changes described in this section make these different uses more explicit.

### All Solidity types allowed as arguments

CVL 1 had some restrictions on the types of arguments allowed in `methods` block
entries.  For example, it was impossible to add entries for functions that
accept arguments with user-defined types (such as enums and structs).

CVL 2 `methods` block entries may use any Solidity types for arguments and

> return values, except for [function types][sol-fn-types].

Author: ```suggestion
return values, except for [function types][sol-fn-types], which are not supported.
```

[sol-fn-types]: https://docs.soliditylang.org/en/v0.8.17/types.html#function-types

To work around the missing types, CVL 1 allowed users to encode some
user-defined types as primitive types in the `methods` block; these workarounds
are no longer allowed in CVL 2.  For example, consider the following solidity
function:

```solidity
contract Example {
    enum Permission { READ, WRITE };

    function f(Permission p) internal { ... }
}
```

In CVL 1, a methods block entry for `f` would need to declare that it takes a
`uint8` argument:

```cvl
methods {
    f(uint8 permission) => NONDET
}
```

In CVL 2, the methods block entry should use the same type as the Solidity
implementations[^contract-types]:

```cvl
methods {
    function f(Example.Permission p) internal => NONDET;
}
```

> 

Author: ````suggestion
The method can be called from CVL as follows:
```cvl
rule example {
    f(Example.Permission.READ);
}
```

````

```cvl
rule example {
    f(Example.Permission.READ);
}
```

[^contract-types]: There is one place where the types do not need to match exactly:
  `address` and specific contract types (such as `IERC20`) are interchangeable.

(cvl2-visibility)=
### Required `internal` or `external` annotation

> Every methods block entry must be marked either `internal` or `external`.  The

Author: After a second read through, it doesn't seem like there's much confusion in the document so I am marking this resolved

any entry without an `_` will only apply to a single contract!
```

(cvl2-returns)=
### Requirements on `returns`

In CVL 1, the `returns` clause on methods block entries was optional.
CVL 2 has stricter requirements on the declared return types.

Entries that apply to specific contracts (i.e. those that don't use the
`_.f` {ref}`syntax <cvl2-wildcards>`) must include a `returns` clause if the
contract method returns a value.  A specific-contract entry may only omit the
`returns` clause if the contract method does not return a value.

The Prover will report an error if the contract method's return type differs
from the type declared in the `methods` block entry.

```{todo}
Error message
```

Wildcard entries must not declare return types, because they may apply to
multiple methods that return different types.

```{todo}
Error message
```

If a wildcard entry has a ghost or function summary, the user must explicitly
provide an `expect` clause to the summary.  The `expect` clause tells the
Prover how to encode the value returned by the summary.  For example:

```cvl
methods {
    function _.foo() internal => fooImpl() expect uint256 ALL;
}
```

> This entry will replace any call to any internal function `f()` with a call to

Reviewer 5: ```suggestion
This entry will replace any call to any internal function `foo()` with a call to
```
Should this example refer to function `foo` rather than `f` ?

```{todo}
Error message
```

````{warning}
The Prover is unable to check that the return type declared in the `expect`
clause matches the return type that the contract expects.  Continuing the above
example, suppose the contract being verified declared a method `foo()` that
returns a type other than `uint256`:

```solidity
function foo() internal returns(address) {
    ...
}

function bar() internal {
    address x = foo();
}
```

In this case, the Prover would encode the value returned by `fooImpl()` as a
`uint256`, and the `bar` method would then attempt to decode this value as an
`address`.  This will cause undefined behavior, and in some cases the Prover
will not be able to detect the error.
````

Changes to integer types
------------------------

In CVL 1, the rules for casting between integer types were complex; CVL 2
simplifies them.

The general rule of thumb is that you should use `mathint` for all data that is
returned from contract functions, and the appropriate `int` or `uint` type for
all data that will be passed as input to contract functions.

It is now impossible for CVL math expressions to cause overflow - all integer
operations are exact.

> 

Author: ```suggestion
The remainder of this section describes the changes in detail.

```

> ### Mathematical operations return `mathint`

Author: ```suggestion
(cvl2-mathops-return-mathint)=
### Mathematical operations return `mathint`
```

contract functions.

It is now impossible for CVL math expressions to cause overflow - all integer
operations are exact.
The remainder of this section describes the changes in detail.

(cvl2-mathops-return-mathint)=
### Mathematical operations return `mathint`

In CVL 2, the results of all arithmetic operators have type `mathint`,
regardless of the input types.  Arithmetic operators include `+`,
`*`, `-`, `/`, `^`, and `%`, but not bitwise operators like `<<`, `xor`, and `|`
(changes to bitwise operators are described {ref}`below <cvl2-bitwise>`).

The primary impact of this change is that you may need to declare more of your
variables as `mathint` instead of `uint`.  If you are performing arithmetic

> operations and integers that you are passing to specs, you will need to be more

Reviewer 5: I am having trouble parsing this sentence as written. Is it possible that we want to say "...performing arithmetic operations and _generating_ integers..." or something equivalent?

explicit about the overflow behavior by using the {ref}`new casting operators
<cvl2-casting>`.

```{todo}
If you do not change this, you will see the following error:
```

### Comparisons require identical types

When comparing two integers using `==`, `<=`, `<`, `>`, or `>=`, CVL 2 will
require both sides of the equation to have identical types, and {ref}`implicit
casts <cvl2-casting>` will not be used.  Comparisons with number literals (e.g.
`0` or `1`) are allowed for any integer type.

If you do not have identical types, the best solution is to use the special
`to_mathint` operator to convert both sides to `mathint`.  For example:

```cvl
assert to_mathint(balanceOf(user)) == initial + deposit;
```

Note that in this example, we do not need to cast the right hand side, since
the result of `+` is always of type `mathint`.

````{todo}
When should you not simply cast to `mathint`?  We have one example: consider the
following code:

```cvl
ghost uint256 sum;

hook ... {
    havoc sum assuming sum@new == sum@old + newBalance - oldBalance;
}
```

Simply casting to `mathint` will turn overflows into vacuity.

In this particular example, the right solution is to declare `sum` to be a

> `mathint` instead of a `uint`.  Note that with the more "modern" update syntax,

Reviewer 5: Can we assume readers will understand what we mean here by "modern" ?

> Simply casting to `mathint` will turn overflows into vacuity.  It's not clear
> how to generalize from this example.
> 
> In this particular example, the right solution is to declare `sum` to be a
> `mathint` instead of a `uint`.  Note that with the more "modern" update syntax,
> this isn't a problem:
> 
> ```cvl
> ghost uint256 sum;
> 
> hook ... {
>     sum = sum + newBalance - oldBalance;
> }
> ```
> will say that the right-hand side is a `mathint` which can't be assigned to a
> `uint`.
> 
> We hope that specification writers will think carefully about these changes and let
> us know of any other situations where the right thing is not simply casting to
> `mathint`.

Author: ```suggestion
Simply casting to `mathint` will turn overflows into vacuity.

In this particular example, the right solution is to declare `sum` to be a
`mathint` instead of a `uint`.  Note that with the more "modern" update syntax,
this isn't a problem:

```cvl
ghost uint256 sum;

hook ... {
    sum = sum + newBalance - oldBalance;
}
```
will say that the right-hand side is a `mathint` which can't be assigned to a
`uint`.

We hope that specification writers will think carefully about these changes and let
us know of any other situations where the right thing is not simply casting to
`mathint`; it is not currently clear exactly how to generalize this example.
```

````

```{todo}
If you do not change this, you will see the following error:
```

(cvl2-casting)=
### Implicit and explicit casting

If every number that can be represented by one type can also be represented by
another type, then we say that the first type is a *subtype* of the second type.

For example, a `uint8` variable could have any value between `0` and `2^8-1`,
and all of these values can be stored in a `uint16` variable, so `uint8` is a
subtype of `uint16`.  An `int16` can also store any value between `0` and
`2^8-1`, so `uint8` is also a subtype of `int16`.

All integer types are subtypes of `mathint`, since any integer can be
represented by a `mathint`.

In CVL 1, the rules for converting between supertypes and subtypes were
complicated; they depended not only on the types involved, but on the context
in which the conversion happened.  CVL 2 simplifies these rules and improves the
clarity and predictability of casts.

In CVL 2, with one exception, you can always use a subtype whenever the
supertype is accepted.  For example, you can always use a `uint8` where an
`int16` is expected.  We say that the subtype can be "implicitly cast" to the
supertype.

> The one exception is comparison operators; as mentioned above, you must add an

Author: ```suggestion
The one exception is comparison operators; as mentioned {ref}`above <cvl2-mathops-return-mathint>`, you must add an
```

explicit conversion if you want to compare two numbers with different types.
The `to_mathint` operator exists solely for this purpose; in all other contexts
you can simply use any number when a `mathint` is expected (since all integer
types are subtypes of `mathint`).

In order to convert from a supertype to a subtype, you must use an explicit
cast.  In CVL 1, only a few casting operators (such as `to_uint256`) were
supported.

> CVL 2 replaces these casting operators with two new casting operators: `assert_<type>(value)` and
> `require_<type>(value)` (for example: `assert_uint8(x)` or `require_uint8(x)`).

Author: ```suggestion
CVL 2 replaces these casting operators with two new casting operators: *assert casts*
such as `assert_uint8(x)` or `assert_int256(x)`, and *require casts* such as `require_uint8(x)` or `require_int256(x)`.
```

> Each of these cases checks that the value is in range; the `assert` cast will

Author: ```suggestion
Each of these casts checks that the value is in range; the `assert` cast will
```

```{note}
By default, bitwise operators are {term}`overapproximated <overapproximation>`,
so you may see counterexamples that incorrectly compute the results of bitwise
operations.

The {ref}`-useBitVectorTheory` flag makes the Prover's reasoning about bitwise
operations more precise, but this flag is experimental in CVL 2.
```

Removed features
----------------

As part of the transition to CVL 2.0, we have removed several language features
that are no longer used.

We have removed these features because we think they are no longer used and no
longer useful.  If you find that you do need one of these features, contact
Certora support.

### Methods entries for sighashes

In CVL 1, you could write a sighash instead of a method identifier in the
`methods` block.  This feature is no longer supported.  You will need to have
the name and argument types of the called method in order to provide an entry.

```{todo}
If you do not change this, you will see the following error:
```

### `invoke`, `sinvoke`, and `call`

Older versions of CVL had special syntax for calling contract and CVL functions:
 - `invoke f(args);` should be replaced with `f(args);`.

>  - `sinvoke f(args);` should be replaced with `f@withrevert(args);`

Reviewer 5: ```suggestion
 - `sinvoke f(args);` should be replaced with `f@withrevert(args);`.
```

>  - `call f(args)` should be replaced with `f(args)`.

Reviewer 5: ```suggestion
 - `call f(args)` should be replaced with `f(args);`.
```
Was the omission of the semicolon correct?

Author: yes, I'm fairly certain `call` is an expression rather than a statement

```{todo}
If you do not change this, you will see the following error:
```

### `static_assert` and `static_require`

These deprecated aliases for `assert` and `require` are being removed; replace
them with `assert` and `require` respectively.

```{todo}
If you do not change this, you will see the following error:
```

### `invoke_fallback`

The `invoke_fallback` syntax is no longer supported; there is no longer a way
to directly invoke the fallback method.

```{todo}
If you do not change this, you will see the following error:
```

### Havocing local variables

In CVL 1, you could write the following:

```cvl
calldataarg args; env e;
f(e, args);

havoc args;
g(e, args);
```

In CVL 2, you can no longer `havoc` local variables.  Instead, replace the

> havoced variable with a new variable.

Reviewer 5: Does it make sense to add:

"So the above example would become..."

?

--------------------------

There are several simple changes to the syntax to make specs more uniform and
consistent, and to reduce the superficial differences with Solidity.

### `function` and `;` required for methods block entries

In CVL 2, methods block entries must now start with `function` and end with
`;` (semicolons were optional in CVL 1).  For example:

```cvl
balanceOf(address) returns(uint) envfree
```
will become
```cvl
function balanceOf(address) external returns(uint) envfree;
```
(note also the addition of `external`, {ref}`described below <cvl2-visibility>`).

This is also true for entries with summaries:
```cvl
_setManagedBalance(address,uint256) => NONDET
```
will become
```cvl
function _setManagedBalance(address,uint256) internal => NONDET;
```

```{todo}
If you do not change this, you will see the following error:
```

### Required `;` in more places

`using`, `pragma`, `import`, and `use` statements all require a `;` at the end.  For
example,

```cvl
using C as c
```

becomes
```cvl
using C as c;
```

`use` statements do not require (and may not have) a semicolon if they
are followed by a `preserved` or `filtered` block.  For example:

```cvl

> use rule tree_isImmutable filtered {
>     f -> !excludeFromProver(f)
> }

Reviewer 5: ```suggestion
use rule poolSolvency filtered {
    f -> !isEmergencyWithdrawal(f)
}
```

```solidity
contract Example {
    struct S {
        uint firstField;
        uint secondField;
        bool thirdField;
    }

    function f() returns(S) { ... }
    function g() returns(uint, uint) { ... }
}
```

To access the return value of `f` in CVL 1, you could write the following:

```cvl
uint x; uint y; bool z;
x, y, z = f();
```

This syntax is no longer supported.  Instead, you should declare a variable with
the struct type:

```cvl
Example.S result = f();
uint x = result.firstField;
```

Destructuring assignments are still allowed for functions that return multiple
values; the following is valid:

```cvl
uint x; uint y;
x, y = g();
```

```{todo}
If you do not change this, you will see the following error:
```

### `bytes[]` and `string[]`

In CVL 1, you could declare variables of type `string[]` and `bytes[]`.  You can
no longer use these types in CVL.

You can still declare contract methods that use these types in the `methods`
block.  However, you can only call methods that take one of these types as an
argument by passing a `calldataarg` variable, and you cannot access the return
value of a method that returns one of these types.

> ```{todo}
> If you do not change this, you will see the following error:
> ```
> 

Reviewer 6: ```suggestion
```
If you do not change this, you will see the following error:
```
An array with an element of type bytes is currently unsupported in CVL. Only primitive elements which fit into one word are allowed.
An array with an element of type string is currently unsupported in CVL. Only primitive elements which fit into one word are allowed.
```

```

Entries in the `methods` block may use either the contract name or the instance
variable:

```cvl
using Example as c;

methods {
    //// both are valid:
    function c.balanceOf(address) external returns(uint) envfree;
    function Example.transfer(address,uint) external envfree;

> }

Reviewer 2: Maybe it is worth it to mention somewhere that the semantics of both function entries are the same for now, but may diverge in the future once contract instances are no longer singleton?

```

% ```{todo}
% Error message
% ```

Changes to methods block entries
--------------------------------

In addition to the superficial changes listed above, there are some changes to
the way that methods block entries can be written (there are also a
{ref}`few instances <cvl2-wildcards>` where the meanings of entries has
changed).  In CVL 1, `methods` block entries often had several different
functions and meanings:

 - They were used to indicate targets for summarization
 - They were used to write generic specs that could apply to contracts with
   missing methods
 - They were used to declare targets `envfree`

The changes described in this section make these different uses more explicit:

```{contents}

> :local:

Reviewer 2: :eyes: fancy!

}
```

Contract functions that take or return contract or interface types should
instead use `address` in the `methods` block declaration.  For example, if the
contract contains the following function:

```solidity
function listToken(IERC20 token) internal
```

the `methods` block should use `address` for the `token` argument:

```cvl
methods {

>     function listToken(IERC20 address) internal

Reviewer 2: this still has type `IERC20`, oopsie

`methods` block.  This feature is no longer supported.  You will need to have
the name and argument types of the called method in order to provide an entry.

% ```{todo}
% If you do not change this, you will see the following error:
% ```

### `invoke`, `sinvoke`, and `call`

Older versions of CVL had special syntax for calling contract and CVL functions:
 - `invoke f(args);` should be replaced with `f@withrevert(args);`.

>  - `sinvoke f(args);` should be replaced with `f(args);`.

Reviewer 2: take it or leave it "should be replaced with `f(args);` _or_ `f@norevert(args);`

Author: I'll leave it; I can't think of any reason to recommend `@norevert`.
